### PR TITLE
Split test atomgroup

### LINKED ISSUE #466: organization of test files
Would it be ok to split up the test-files? Currently some of them are really huge (several thousand lines) and they just keep growing with random additions. I find it really hard to get an overview in these large files. I also have a bad overview which files from the package are tested well. As an example take `test_coordinates.py`. it contains  a lot of doublicated code. It's hard to tell just by looking at the file if a specific reader is tested well or not.

I personally would like a structure like this

```
MDAnalysisTests
   | -- coordinates
             base.py
             test_pdb.py
             test_dcd.py
             test_xdr.py
             ...
```

In such a file structure I find it easier to find tests for specific modules. Individual files aren't monolithic and just by the line-count one can get an impression how well tested individual modules are. 

I'm asking because I will rewrite some of the test_coordinates any way now. I'm doing this because @richardjgowers pointed out in #441 that my xtc implementation is set to the wrong frame after initialization and no test catches that. To me this seems like a test that every reader should have and pass. So I want to write some generic reader/writer test-classes that will make it easier to add such tests fast for all readers. At the some time I would also address #440 for the coordinates tests.
## Files to split up
- [x] test_coordinates.py [#469]
- [ ] test_atomgroup.py
- [x] test_analysis.py [#492]

### LINKED ISSUE #940: attrbute settings aren't propagated up through levels
### Expected behaviour

Settings a resname from the AtomGroup level also changes resnames and residue types in the residue and segment level.

In general I would expect that changes set at the Atom level also travel up to residue and segment groups as vice versa (that works now already)
### Actual behaviour

only residue names are changes but the residueGroup associated with an AtomGroup is still the same
### Code to reproduce the behaviour

``` python
import MDAnalysis as mda

u = mda.Universe(top, trj)
u.atoms.resnames = 'ALA'
print(np.unique(u.atoms.residues.resnames))
# shows more then just one residue type
print(np.unique(u.atoms.resnames))
# only prints 'ALA' so information about the universe has diverged.

# however
u.atoms.residues.set_resname('ALA')
print(np.unique(u.atoms.residues.resnames))
# only prints 'ALA'
```
### Currently version of MDAnalysis:

0.15.1-dev

### LINKED ISSUE #1178: Setting attributes to wrong size currently works.
### Expected behaviour

```
ag = u.atoms[:3]

ag.masses = [1, 2, 3, 4]

# should raise an Error as we're assigning 4 values to 3 atoms
```

### Actual behaviour

Numpy just raises a warning that this won't work in the future.  It should really get to the numpy level through.

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #466 
Fixes #1178
Fixes #940

Changes made
 * `test_atomgroup` split into many smaller files
 * `AtomGroup[atomname]` access reimplemented (eg `u.atoms['HT1']`)
 * size checks for topology attribute assigning
 * populated `MDAnalysisTests.util` from `__init__`
 * moved `make_Universe` to top test namespace
 * added better error messages to setting attributes from wrong level

### CODE REVIEW

File: package/MDAnalysis/core/topologyattrs.py [hunk 1]
```diff
@@ -33,20 +33,70 @@
...
+    set_X(self, group, values)
+
+    if group in (Atom, Residue, Segment):
+        values must be single values, ie int, float or string
+    else:
+        values must be single value OR same length as group
+
+    """
+    _SINGLE_VALUE_ERROR = ("Setting {cls} {attrname} with wrong sized input. "
+                           "Must use single value, length of suppled values: {lenvalues}.")
```
Comment (jbarnoud) [inline package/MDAnalysis/core/topologyattrs.py]: 'supplied' instead of 'suppled'

**Comment by richardjgowers** [comment]
Ok this is good to go hopefully

File: package/MDAnalysis/core/groups.py [hunk 2]
```diff
@@ -975,6 +980,15 @@ class AtomGroup(GroupBase):
     .. SeeAlso:: :class:`MDAnalysis.core.universe.Universe`
 
     """
+    def __getitem__(self, item):
+        # u.atoms['HT1'] access, otherwise default
+        if isinstance(item, string_types):
```
Comment (jbarnoud) [inline package/MDAnalysis/core/groups.py]: I do not see this behavior in the docstring above. Is this documented somewhere?

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -32,21 +32,123 @@
...
+
+    Eg:
+    setting mass of residue, gets called with attr=Masses, group=residue
+
+    raises a NotImplementedError with:
+    'Cannot set masses from Residue.  Use 'Residue.atoms.masses'
+
+    Mainly used to ensure consistent and helpful error messages
+    """
+    err_msg = "Cannot set {attr} from {cls}. "
```
Comment (jbarnoud) [inline package/MDAnalysis/core/topologyattrs.py]: I think I would prefer this message template to be down in the function, closer to where it is used. But it really is a matter of taste.

File: package/MDAnalysis/core/groups.py (hunk 2 above)
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: https://github.com/MDAnalysis/mdanalysis/blob/master/package/MDAnalysis/core/AtomGroup.py#L3524

It was in the previous incarnation of docs, so it's part of the tech debt in #1078

File: testsuite/MDAnalysisTests/core/test_atom.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    def test_add_AG(self):
+        ag = self.universe.atoms[:2]
+
+        ag2 = self.atom + ag
+
+        for at in [self.atom, ag[0], ag[1]]:
+            assert_equal(at in ag2, True)
+
+    def test_no_velo(self):
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atom.py]: I guess this test just got moved from somewhere else. But isn't it redundant with `test_velocity_fail` bellow?

File: package/MDAnalysis/core/topologyattrs.py (hunk 3 above)
Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: Im' also for defining variables where they are used.

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 5]
```diff
@@ -190,6 +198,117 @@ def test_bogus_kwarg_pdb(self):
...
+                     "written s4AKE segment does not match original segment")
+        assert_almost_equal(
+            G2.positions, G.positions, self.precision,
+            err_msg="segment s4AKE coordinates do not agree with original")
+
+    def test_write_Universe(self):
+        U = self.universe
+        W = mda.Writer(self.outfile)
+        W.write(U)
+        W.close()
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: Once again, probably an old test. Could use `with`, though, to avoid hanging files if the test fails.

File: testsuite/MDAnalysisTests/core/test_atom.py (hunk 4 above)
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atom.py]: https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/core/test_group_traj_access.py#L247

It's very redundant because we systematically go through all possible combinations there.  But they've got some values as these use files, so are more "real"?

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 6]
```diff
@@ -393,3 +512,779 @@ def test_split_VE(self):
         ag = self.universe.atoms[:40]
 
         assert_raises(ValueError, ag.split, 'something')
+
+
+class TestWrap(object):
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I was not aware of `AtomGroup.wrap`. Great!

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 7]
```diff
@@ -393,3 +512,779 @@ def test_split_VE(self):
...
+                ('type', 'types', 'string'),
+                ('altLoc', 'altLocs', 'string'),
+                ('charge', 'charges', 'float'),
+                ('mass', 'masses', 'float'),
+                ('radius', 'radii', 'float'),
+                ('bfactor', 'bfactors', 'float'),
+                ('occupancy', 'occupancies', 'float')
+        ):
+            vals = self.get_new(att_type)
+            yield self._check_ag_matches_atom, att, atts, ag
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: Whoever makes the transition to py.test will hate this :tongue:

Comment (kain88-de) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I dislike this already. I find the yield test generators hard to follow in nose in general.

Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I really like them and I will miss them a lot. I like how they allow you to build you series of tests so as the test itself only contains the minimum, all the hand waving is done at building time.

Comment (kain88-de) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I can do that in pytest as well. Just a hell of a lot more readable.  For [example](https://github.com/datreant/datreant.core/blob/44c7664cd5c3889514080e44c5f3fc64a615509c/src/datreant/core/tests/test_treants.py#L169). This is a marked fixture attached to the function directly, all values the second parameter will be passed to test after the discovery phase. The equivalent with nose-yield would be less readable IMHO.

Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I now about the `parametrized` fixture, but taking the lines we are discussing under right now, how would you have the line `vals = self.get_new(att_type)` out of the tests using `parametrized`?

Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: I think you'd just have to put the `vals = ` line into the test method.  I agree that the py.test syntax is nicer!

Comment (jbarnoud) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py]: In some cases, the `vals = ` equivalent part can be larger and I think it should not be part of the test. I agree, though, that the py.test syntax *looks* nicer.

**Comment by jbarnoud** [comment]
Is anybody planning on having a look? If not I merge in a few hours.

File: package/MDAnalysis/core/topologyattrs.py [hunk 8]
```diff
@@ -32,21 +32,122 @@
 """
 
 from six.moves import zip, range
+
+import Bio.Seq
+import Bio.SeqRecord
+import Bio.Alphabet
```
Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: Where/Why do we need the Biopython imports? I can't find anything in the diff.

File: package/MDAnalysis/core/topologyattrs.py [hunk 9]
```diff
@@ -32,21 +32,122 @@
...
+        # quasi len measurement
+        # strings, floats, ints are len 0, ie not iterable
+        # other iterables are just len'd
+        if iterable(values):
+            return len(values)
+        else:
+            return 0  # special case
+
+    @functools.wraps(func)
+    def wrapper(attr, group, values):
```
Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: where is this used? I can't find it in the diff for this file.

Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: Also a doc-string or a name for what the wrapper is thought to be used would be nice.

File: package/MDAnalysis/core/topologyattrs.py (hunk 8 above)
Comment (jbarnoud) [inline package/MDAnalysis/core/topologyattrs.py]: Biopython used to be imported inside of a class (https://github.com/MDAnalysis/mdanalysis/pull/1164/files#diff-96c7f631815d065c6f087f4aea01a4e6L1035). The PR is just moving the imports at the top.

File: package/MDAnalysis/core/topologyattrs.py (hunk 9 above)
Comment (richardjgowers) [inline package/MDAnalysis/core/topologyattrs.py]: This is inside the `_check_length` wrapper above, it's documented there

**Comment by richardjgowers** [comment]
Ok I'm done fiddling with this

**Comment by richardjgowers** [comment]
thanks @jbarnoud & @kain88-de !
